File: packages/core/src/recurring.ts
import { PERSONAL, blake2b } from "@nervosnetwork/ckb-sdk-utils";

import {
  parseBlockNumber,
  parseRunCount,
  parseShannons,
  type BlockNumber,
  type RunCount,
  type Shannons,
} from "./chain-values.ts";

export const POLICY_PAYLOAD_DOMAIN = "ckb-automata/policy-payload/v1" as const;

const BYTE32_LENGTH = 32;
const MAX_UINT32 = 0xffff_ffff;
const MAX_UINT32_BIGINT = 0xffff_ffffn;
const MAX_ABSOLUTE_BLOCK_NUMBER = (1n << 56n) - 1n;

function requireByte32(value: Uint8Array, name: string): Uint8Array {
  if (value.length !== BYTE32_LENGTH) {
    throw new RangeError(`${name} must contain exactly 32 bytes`);
  }
  return value;
}

export function deriveRecurringPayloadHash(
  policyScriptHash: Uint8Array,
  payload: Uint8Array,
): Uint8Array {
  const policyHash = requireByte32(policyScriptHash, "policyScriptHash");
  const bodyLength = policyHash.length + payload.length;
  if (bodyLength > MAX_UINT32) {
    throw new RangeError("recurring payload commitment exceeds uint32 length");
  }
  const encodedLength = new Uint8Array(4);
  new DataView(encodedLength.buffer).setUint32(0, bodyLength, true);

  const hasher = blake2b(32, null, null, PERSONAL);
  hasher.update(new TextEncoder().encode(POLICY_PAYLOAD_DOMAIN));
  hasher.update(Uint8Array.of(0));
  hasher.update(encodedLength);
  hasher.update(policyHash);
  hasher.update(payload);
  return hasher.digest("binary") as Uint8Array;
}

export interface RecurringSchedule {
  readonly amount: Shannons;
  readonly intervalBlocks: BlockNumber;
  readonly firstNotBefore: BlockNumber;
  readonly totalRuns: RunCount;
  readonly finalRefundKind: number;
}

export function isValidRecurringSchedule(schedule: RecurringSchedule): boolean {
  try {
    const amount = parseShannons(schedule.amount);
    const intervalBlocks = parseBlockNumber(schedule.intervalBlocks);
    const firstNotBefore = parseBlockNumber(schedule.firstNotBefore);
    const totalRuns = parseRunCount(schedule.totalRuns);
    return (
      amount > 0n &&
      intervalBlocks > 0n &&
      firstNotBefore > 0n &&
      firstNotBefore <= MAX_ABSOLUTE_BLOCK_NUMBER &&
      totalRuns > 0n &&
      totalRuns <= MAX_UINT32_BIGINT &&
      schedule.finalRefundKind === 0
    );
  } catch {
    return false;
  }
}

File: contracts/shared/recurring.rs
use ckb_hash::new_blake2b;

const POLICY_PAYLOAD_DOMAIN: &[u8] = b"ckb-automata/policy-payload/v1";
const MAX_ABSOLUTE_BLOCK_NUMBER: u64 = (1 << 56) - 1;

pub fn recurring_payload_hash(
    policy_script_hash: &[u8; 32],
    payload: &[u8],
) -> Option<[u8; 32]> {
    let body_length = policy_script_hash
        .len()
        .checked_add(payload.len())
        .and_then(|length| u32::try_from(length).ok())?;
    let mut hasher = new_blake2b();
    hasher.update(POLICY_PAYLOAD_DOMAIN);
    hasher.update(&[0]);
    hasher.update(&body_length.to_le_bytes());
    hasher.update(policy_script_hash);
    hasher.update(payload);
    let mut result = [0_u8; 32];
    hasher.finalize(&mut result);
    Some(result)
}

pub fn valid_recurring_schedule(
    amount: u64,
    interval_blocks: u64,
    first_not_before: u64,
    total_runs: u32,
    final_refund_kind: u8,
) -> bool {
    amount > 0
        && interval_blocks > 0
        && first_not_before > 0
        && first_not_before <= MAX_ABSOLUTE_BLOCK_NUMBER
        && total_runs > 0
        && final_refund_kind == 0
}
